logdna/resource_category_test.go

```diff
@@ -0,0 +1,90 @@
+package logdna
+
+import (
+  "fmt"
+  "regexp"
+  "testing"
+
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestCategory_ErrorProviderUrl(t *testing.T) {
+  pcArgs := []string{serviceKey, "https://api.logdna.co"}
+  catArgs := map[string]string{
+    "name": `"test-category"`,
+    "type": `"views"`,
+  }
+
+  resource.Test(t, resource.TestCase{
+    Providers: testAccProviders,
+    Steps: []resource.TestStep{
+      {
+        Config: fmtTestConfigResource("category", "new", pcArgs, catArgs, nilOpt, nilLst),
+        ExpectError: regexp.MustCompile("Error: error during HTTP request: Post \"https://api.logdna.co/v1/config/categories/views\": dial tcp: lookup api.logdna.co"),
+      },
+    },
+  })
+}
+
+func TestCategory_ErrorResourceName(t *testing.T) {
+  catArgs := map[string]string{
+    "type": `"views"`,
+  }
+
+  resource.Test(t, resource.TestCase{
+    Providers: testAccProviders,
+    Steps: []resource.TestStep{
+      {
+        Config: fmtTestConfigResource("category", "new", nilLst, catArgs, nilOpt, nilLst),
+        ExpectError: regexp.MustCompile("The argument \"name\" is required, but no definition was found."),
+      },
+    },
+  })
+}
+
+func TestCategory_Basic(t *testing.T) {
```

Comment:
it looks like we're missing a test case for an update here. 

here's one example but there's other too https://github.com/logdna/terraform-provider-logdna/blob/c87a6cc7be8fbf70ee2531923580702d08dc80e7/logdna/resource_alert_test.go#L226-L230

iirc the `delete` will be tested implicitly by the testing framework (it automatically destroys resources after each test by default)

Reply:
@mdeltito @jorgebay The name of both test cases is the same, the first case tests create action and the second one tests update action. Overall this case tests the entire lifecycle of a CRUD interface.
<img width="1504" alt="Screen Shot 2022-04-19 at 11 21 42 PM" src="https://user-images.githubusercontent.com/5346824/164144630-ee139396-ad58-4eaf-a590-b58dd5e98f45.png">
<img width="364" alt="Screen Shot 2022-04-19 at 11 23 05 PM" src="https://user-images.githubusercontent.com/5346824/164144686-37572ab1-1f53-45b0-8abf-dc4f61e4dfb5.png">

Reply:
Got it, @seeruyy, thanks.  I do think there's a few style things we could do better on which may help clear up confusion.  Couple things here:

* It's a terrible pattern that we've started using `new` for the resource names.  That is a reserved word in some languages and just doesn't describe the resource.  Even something like `new_thing` would be better.  I'd like to start naming things for what they are and break that pattern (no, we don't have to retroactively fix things as long as new code does better).

* We should rely on heavily using `t.Run` ([like this](https://github.com/logdna/terraform-provider-logdna/blob/529d2fa80335b854fa4aefa67a46cb04bb582851/logdna/request_test.go#L41-L46)) to give proper descriptions and subtests.  The confusion on this test could be cleared up by 2 simple subtests--the first with a description of "creating blah blah" and the second of "updating the same blah blah".

These seem like nits, but we'd do ourselves a favor by improving just a bit so that we don't hurt the readability and intuitiveness of the code.  The tests have already gotten pretty abstracted and can be hard to follow, IMO.

Reply:
> We should rely on heavily using t.Run ([like this](https://github.com/logdna/terraform-provider-logdna/blob/529d2fa80335b854fa4aefa67a46cb04bb582851/logdna/request_test.go#L41-L46)) to give proper descriptions and subtests

I don't think that is true for tests that are using the terraform sdk/test suite? I think the confusion is just that there's some "magic" going on via the sdk which makes it unclear (especially upon picking this project up again) whether we're missing assertions or not.

Reply:
Sorry, I didn't mean to start a big debate.  Yes, there's magic with that test, but it seems that it could be easily wrapped within a child test with a description.  Not having descriptions in the test title or in the various assertions is sucky, IMO.